be/src/vec/functions/function_java_udf.cpp

```diff
@@ -130,189 +141,177 @@ Status JavaFunctionCall::execute(FunctionContext* context, Block& block,
+            arr_obj = (jobjectArray)env->CallNonvirtualObjectMethod(
+                    jni_ctx->executor, jni_env->executor_cl,
+                    jni_env->executor_convert_basic_argument_id, arg_idx, arg_column_nullable,
+                    num_rows, nullmap_address,
+                    reinterpret_cast<int64_t>(data_cols[arg_idx]->get_raw_data().data), 0);
+        } else if (data_cols[arg_idx]->is_column_string()) {
             const ColumnString* str_col =
                     assert_cast<const ColumnString*>(data_cols[arg_idx].get());
-            jni_ctx->input_values_buffer_ptr.get()[arg_idx] =
-                    reinterpret_cast<int64_t>(str_col->get_chars().data());
-            jni_ctx->input_offsets_ptrs.get()[arg_idx] =
-                    reinterpret_cast<int64_t>(str_col->get_offsets().data());
-        } else if (data_cols[arg_idx]->is_numeric() || data_cols[arg_idx]->is_column_decimal()) {
-            jni_ctx->input_values_buffer_ptr.get()[arg_idx] =
-                    reinterpret_cast<int64_t>(data_cols[arg_idx]->get_raw_data().data);
+            arr_obj = (jobjectArray)env->CallNonvirtualObjectMethod(
+                    jni_ctx->executor, jni_env->executor_cl,
+                    jni_env->executor_convert_basic_argument_id, arg_idx, arg_column_nullable,
+                    num_rows, nullmap_address,
+                    reinterpret_cast<int64_t>(str_col->get_chars().data()),
+                    reinterpret_cast<int64_t>(str_col->get_offsets().data()));
         } else if (data_cols[arg_idx]->is_column_array()) {
             const ColumnArray* array_col =
                     assert_cast<const ColumnArray*>(data_cols[arg_idx].get());
-            jni_ctx->input_offsets_ptrs.get()[arg_idx] =
-                    reinterpret_cast<int64_t>(array_col->get_offsets_column().get_raw_data().data);
             const ColumnNullable& array_nested_nullable =
                     assert_cast<const ColumnNullable&>(array_col->get_data());
             auto data_column_null_map = array_nested_nullable.get_null_map_column_ptr();
             auto data_column = array_nested_nullable.get_nested_column_ptr();
-            jni_ctx->input_array_nulls_buffer_ptr.get()[arg_idx] = reinterpret_cast<int64_t>(
+            auto offset_address =
+                    reinterpret_cast<int64_t>(array_col->get_offsets_column().get_raw_data().data);
+            auto nested_nullmap_address = reinterpret_cast<int64_t>(
                     check_and_get_column<ColumnVector<UInt8>>(data_column_null_map)
                             ->get_data()
                             .data());
-
-            //need pass FE, nullamp and offset, chars
+            int64_t nested_data_address = 0, nested_offset_address = 0;
+            // array type need pass address: [nullmap_address], offset_address, nested_nullmap_address, nested_data_address/nested_char_address,nested_offset_address
             if (data_column->is_column_string()) {
```

Comment:
Could we use Macro to collapse this pass above?

Reply:
seems could use macro, but it will reduce the readability of the code and maybe not be worth it.